File: ui/src/api/userSettings.ts
import { callMcpTool } from "@/api/mcpClient";

/**
 * Per-user, per-ROLE ordered model selection ("lanes"). Each lane is an ordered
 * fallback list (priority high → low) of full `"provider/model"` ids. A task's
 * base role maps to one lane: `plan` (planner/architect/chat), `implement`
 * (worker), `review` (reviewer).
 */
export interface ModelLanes {
  plan: string[];
  implement: string[];
  review: string[];
}

/** Per-user concurrency ceilings for each role lane. */
export interface LaneMaxSessions {
  plan: number;
  implement: number;
  review: number;
}

/** The three lane keys, in display order. */
export const MODEL_LANE_KEYS = ["plan", "implement", "review"] as const;
export type ModelLaneKey = (typeof MODEL_LANE_KEYS)[number];

export function emptyLanes(): ModelLanes {
  return { plan: [], implement: [], review: [] };
}

/** Conservative defaults used when a user explicitly configures lane limits. */
export function defaultLaneMaxSessions(): LaneMaxSessions {
  return { plan: 1, implement: 1, review: 1 };
}

/**
 * Distinct union of all model ids across every lane (order: plan, implement,
 * review; duplicates dropped). Used where a single flat per-user selection is
 * still meaningful — e.g. the chat model picker and the onboarding gate.
 */
export function lanesUnion(lanes: ModelLanes): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const key of MODEL_LANE_KEYS) {
    for (const id of lanes[key]) {
      if (!seen.has(id)) {
        seen.add(id);
        out.push(id);
      }
    }
  }
  return out;
}

/**
 * Minimum per-model cap needed for all configured lane ceilings to be usable
 * simultaneously. A model selected in multiple lanes needs the sum because
 * its model cap is shared across those lanes.
 */
export function aggregateModelCapacity(
  lanes: ModelLanes,
  laneMaxSessions: LaneMaxSessions,
): Record<string, number> {
  const aggregate: Record<string, number> = {};
  for (const lane of MODEL_LANE_KEYS) {
    for (const modelId of new Set(lanes[lane])) {
      aggregate[modelId] = (aggregate[modelId] ?? 0) + laneMaxSessions[lane];
    }
  }
  return aggregate;
}

/** Normalise a raw, possibly-partial lanes object into a complete `ModelLanes`. */
export function parseLanes(raw: Partial<ModelLanes> | null | undefined): ModelLanes {
  return {
    plan: Array.isArray(raw?.plan) ? raw.plan : [],
    implement: Array.isArray(raw?.implement) ? raw.implement : [],
    review: Array.isArray(raw?.review) ? raw.review : [],
  };
}

/**
 * Parse an optional wire-level lane limit without erasing the legacy/unset
 * distinction. The server returns all three fields together; malformed or
 * partial payloads are treated as unset instead of inventing missing limits.
 */
export function parseLaneMaxSessions(raw: unknown): LaneMaxSessions | undefined {
  if (!raw || typeof raw !== "object") return undefined;
  const candidate = raw as Partial<Record<ModelLaneKey, unknown>>;
  const values = MODEL_LANE_KEYS.map((lane) => candidate[lane]);
  if (
    values.some(
      (value) =>
        !Number.isInteger(value) ||
        (value as number) < 1 ||
        (value as number) > 10,
    )
  ) {
    return undefined;
  }
  return {
    plan: candidate.plan as number,
    implement: candidate.implement as number,
    review: candidate.review as number,
  };
}

export interface UserSettings {
  autoApprovePrs: boolean;
  /** Per-role model lanes (priority high → low per lane). */
  lanes: ModelLanes;
  /**
   * Per-user per-model concurrency caps, keyed by full `"provider/model"` id.
   * `{}` when unset; consumers default missing entries to 1.
   */
  maxSessions: Record<string, number>;
  /** Per-lane concurrency ceilings; undefined preserves legacy/unbounded users. */
  laneMaxSessions?: LaneMaxSessions;
  /**
   * Cross-model ("Thorough") review. When true (the default), the reviewer
   * prefers a model id different from the implementer's. A degenerate
   * single-model selection falls back to same-model review.
   */
  diverseReview: boolean;
  /**
   * Cross-model ("Diverse") refinement. When true (the default), the
   * proposal-refinement roles (advocate, adversary, judge) prefer a model id
   * different from the primary task model. Falls back to same-model when
   * alternatives are unavailable.
   */
  diverseRefinement: boolean;
}

interface RawGet {
  ok?: boolean;
  user_id?: string | null;
  auto_approve_prs?: boolean;
  lanes?: Partial<ModelLanes> | null;
  max_sessions?: Record<string, number> | null;
  lane_max_sessions?: LaneMaxSessions | null;
  diverse_review?: boolean | null;
  diverse_refinement?: boolean | null;
  error?: string | null;
}

interface RawSet {
  ok?: boolean;
  applied?: boolean;
  auto_approve_prs?: boolean | null;
  lanes?: Partial<ModelLanes> | null;
  max_sessions?: Record<string, number> | null;
  lane_max_sessions?: LaneMaxSessions | null;
  diverse_review?: boolean | null;
  diverse_refinement?: boolean | null;
  error?: string | null;
}

function parseMaxSessions(raw: Record<string, number> | null | undefined): Record<string, number> {
  return raw && typeof raw === "object" ? raw : {};
}

export async function fetchUserSettings(): Promise<UserSettings> {
  const resp = (await callMcpTool("user_settings_get", {})) as RawGet;
  if (resp?.ok === false) {
    throw new Error(resp.error ?? "failed to load user settings");
  }
  return {
    autoApprovePrs: Boolean(resp?.auto_approve_prs),
    lanes: parseLanes(resp?.lanes),
    maxSessions: parseMaxSessions(resp?.max_sessions),
    laneMaxSessions: parseLaneMaxSessions(resp?.lane_max_sessions),
    // Cross-model review defaults ON; only an explicit `false` disables it.
    diverseReview: resp?.diverse_review !== false,
    // Cross-model refinement defaults ON; only an explicit `false` disables it.
    diverseRefinement: resp?.diverse_refinement !== false,
  };
}

export async function patchUserSettings(patch: {
  autoApprovePrs?: boolean;
  /** Per-role model lanes (priority order per lane). Omit to keep current. */
  lanes?: ModelLanes;
  /** Per-model caps keyed by full `"provider/model"` id. Omit to keep current. */
  maxSessions?: Record<string, number>;
  /** Per-role concurrency ceilings. Omit to keep current/unset. */
  laneMaxSessions?: LaneMaxSessions;
  /** Cross-model ("Thorough") review toggle. Omit to keep current. */
  diverseReview?: boolean;
  /** Cross-model ("Diverse") refinement toggle. Omit to keep current. */
  diverseRefinement?: boolean;
}): Promise<UserSettings> {
  const args: Record<string, unknown> = {};
  if (patch.autoApprovePrs !== undefined) {
    args.auto_approve_prs = patch.autoApprovePrs;
  }
  if (patch.lanes !== undefined) {
    args.lanes = patch.lanes;
  }
  if (patch.maxSessions !== undefined) {
    args.max_sessions = patch.maxSessions;
  }
  if (patch.laneMaxSessions !== undefined) {
    args.lane_max_sessions = patch.laneMaxSessions;
  }
  if (patch.diverseReview !== undefined) {
    args.diverse_review = patch.diverseReview;
  }
  if (patch.diverseRefinement !== undefined) {
    args.diverse_refinement = patch.diverseRefinement;
  }
  const resp = (await callMcpTool("user_settings_set", args)) as RawSet;
  if (resp?.ok === false) {
    throw new Error(resp.error ?? "failed to save user settings");
  }
  return {
    autoApprovePrs: Boolean(resp?.auto_approve_prs),
    lanes: parseLanes(resp?.lanes),
    maxSessions: parseMaxSessions(resp?.max_sessions),
    laneMaxSessions: parseLaneMaxSessions(resp?.lane_max_sessions),
    diverseReview: resp?.diverse_review !== false,
    diverseRefinement: resp?.diverse_refinement !== false,
  };
}

File: server/crates/djinn-core/src/models/user_settings.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

/// Per-user, per-ROLE ordered model selection ("lanes"). Each lane is an
/// ordered fallback list (highest priority first) of full `provider/model`
/// ids. A task's base role maps to exactly one lane:
///
/// - `plan`      → planner, architect, chat
/// - `implement` → worker
/// - `review`    → reviewer
///
/// (`lead` is a human role with no model; dispatch fallbacks treat it as
/// `plan`.) Persisted as a single JSON-object TEXT column
/// (`user_settings.model_lanes`, migration 77). An all-empty `ModelLanes`
/// reads back as `None` on `UserSettings.lanes` → global fallback.
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ModelLanes {
    #[serde(default)]
    pub plan: Vec<String>,
    #[serde(default)]
    pub implement: Vec<String>,
    #[serde(default)]
    pub review: Vec<String>,
}

/// Per-user concurrency ceiling for autonomous work in each model lane.
///
/// This is deliberately separate from [`UserSettings::max_sessions`], whose
/// caps are keyed by model and shared across every lane that uses that model.
/// `lane_max_sessions` limits the amount of work admitted for a role even when
/// several lanes happen to use the same model. Interactive chat is not subject
/// to these ceilings even though chat uses the plan lane for model selection.
///
/// Persisted as a JSON-object TEXT column (`user_settings.lane_max_sessions`,
/// migration 114). An absent value means no lane-specific ceiling, preserving
/// the behavior of users created before lane limits were introduced.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct LaneMaxSessions {
    pub plan: u32,
    pub implement: u32,
    pub review: u32,
}

/// The three lanes a per-user model selection is split across.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ModelLane {
    Plan,
    Implement,
    Review,
}

impl ModelLane {
    /// Map a dispatch base role to its lane. `lead` is a human role with no
    /// model; it falls back to the `plan` lane so escalation paths that resolve
    /// a lead model still have a candidate.
    pub fn for_role(base_role: &str) -> Self {
        match base_role {
            "worker" => ModelLane::Implement,
            "reviewer" => ModelLane::Review,
            // planner, architect, chat, lead, and anything unknown → plan
            _ => ModelLane::Plan,
        }
    }
}

impl LaneMaxSessions {
    pub const MIN: u32 = 1;
    pub const MAX: u32 = 10;

    /// The configured ceiling for a lane.
    pub fn lane(&self, lane: ModelLane) -> u32 {
        match lane {
            ModelLane::Plan => self.plan,
            ModelLane::Implement => self.implement,
            ModelLane::Review => self.review,
        }
    }

    /// The configured ceiling for the lane a dispatch role maps to.
    pub fn for_role(&self, base_role: &str) -> u32 {
        self.lane(ModelLane::for_role(base_role))
    }

    /// True when every ceiling is inside the supported persisted/API range.
    pub fn is_valid(&self) -> bool {
        [self.plan, self.implement, self.review]
            .into_iter()
            .all(|value| (Self::MIN..=Self::MAX).contains(&value))
    }
}

impl ModelLanes {
    /// True when every lane is empty (→ treated as "no explicit selection").
    pub fn is_empty(&self) -> bool {
        self.plan.is_empty() && self.implement.is_empty() && self.review.is_empty()
    }

    /// The ordered list for a given lane.
    pub fn lane(&self, lane: ModelLane) -> &[String] {
        match lane {
            ModelLane::Plan => &self.plan,
            ModelLane::Implement => &self.implement,
            ModelLane::Review => &self.review,
        }
    }

    /// The ordered list for the lane a base role maps to.
    pub fn for_role(&self, base_role: &str) -> &[String] {
        self.lane(ModelLane::for_role(base_role))
    }

    /// Distinct union of all model ids across all three lanes (order: plan,
    /// implement, review; duplicates dropped). Used by the slot pool to size
    /// capacity for everything a user might dispatch.
    pub fn all_models(&self) -> Vec<String> {
        let mut seen = std::collections::HashSet::new();
        let mut out = Vec::new();
        for m in self.plan.iter().chain(&self.implement).chain(&self.review) {
            if seen.insert(m.clone()) {
                out.push(m.clone());
            }
        }
        out
    }

    /// Build lanes from a legacy flat list by seeding every lane with the same
    /// ordered selection. Used by the cut-over migration's read path and any
    /// remaining flat-list callers.
    pub fn from_flat(models: Vec<String>) -> Self {
        Self {
            plan: models.clone(),
            implement: models.clone(),
            review: models,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[cfg_attr(feature = "sqlx", derive(sqlx::FromRow))]
pub struct UserSettings {
    pub user_id: String,
    pub auto_approve_prs: bool,
    /// Per-user, per-ROLE ordered model selection ("lanes"). `None` = no
    /// explicit selection → callers fall back to the global deployment model
    /// list. Persisted as a JSON-object TEXT column (`user_settings.model_lanes`,
    /// migration 77).
    #[cfg_attr(feature = "sqlx", sqlx(default))]
    pub lanes: Option<ModelLanes>,
    /// Per-user, per-model concurrency caps (`{ "provider/model": cap }`). The
    /// per-model admission ceiling at dispatch — the slot pool spawns on demand,
    /// with no global ceiling. Composed with `lane_max_sessions` when a user has
    /// lane-specific ceilings. `None`/absent ⇒ default 1 per selected model.
    /// Persisted as a JSON-object TEXT column (`user_settings.max_sessions`,
    /// migration 32). Caps are per-model, shared across lanes.
    #[cfg_attr(feature = "sqlx", sqlx(default))]
    pub max_sessions: Option<HashMap<String, u32>>,
    /// Per-user concurrency ceilings for plan, implement, and review work.
    /// `None` means no lane-specific ceiling (legacy/unbounded behavior).
    /// Persisted as a JSON-object TEXT column
    /// (`user_settings.lane_max_sessions`, migration 114).
    #[cfg_attr(feature = "sqlx", sqlx(default))]
    pub lane_max_sessions: Option<LaneMaxSessions>,
    /// Cross-model ("Thorough") review. When `true` (the default), a task
    /// dispatched to the reviewer role prefers a model id different from the one
    /// that implemented the task — the reviewer walks its review-lane fallback
    /// list and takes the first id that differs from the implementer's. A
    /// degenerate single-model selection silently falls back to same-model.
    /// Persisted as a NOT-NULL boolean column (`user_settings.diverse_review`,
    /// migration 78, default TRUE).
    pub diverse_review: bool,
    /// Cross-model ("Diverse") refinement. When `true` (the default), the
    /// proposal-refinement roles (advocate, adversary, judge) prefer a model id
    /// different from the primary task model — using best-effort cross-model
    /// diversity. When alternatives are unavailable, dispatch collapses to
    /// same-model rather than blocking. Persisted as a NOT-NULL boolean column
    /// (`user_settings.diverse_refinement`, migration 81, default TRUE).
    pub diverse_refinement: bool,
    pub created_at: String,
    pub updated_at: String,
}

impl UserSettings {
    /// All-defaults-off row for users who have never written a setting. Lets
    /// the read path return `Some(defaults)` without inserting a row, so a
    /// `get` call for a freshly-created user doesn't pollute the table.
    pub fn defaults_for(user_id: &str) -> Self {
        Self {
            user_id: user_id.to_string(),
            auto_approve_prs: false,
            lanes: None,
            max_sessions: None,
            lane_max_sessions: None,
            // Cross-model review is on by default (DB column default TRUE), so a
            // never-written user gets the same behavior as a migrated row.
            diverse_review: true,
            // Cross-model refinement is on by default (DB column default TRUE),
            // so a never-written user gets the same behavior as a migrated row.
            diverse_refinement: true,
            created_at: String::new(),
            updated_at: String::new(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lane_max_sessions_maps_dispatch_roles() {
        let limits = LaneMaxSessions {
            plan: 1,
            implement: 3,
            review: 2,
        };

        assert_eq!(limits.for_role("planner"), 1);
        assert_eq!(limits.for_role("architect"), 1);
        assert_eq!(limits.for_role("worker"), 3);
        assert_eq!(limits.for_role("reviewer"), 2);
    }

    #[test]
    fn lane_max_sessions_validates_supported_range() {
        assert!(
            LaneMaxSessions {
                plan: 1,
                implement: 10,
                review: 3,
            }
            .is_valid()
        );
        assert!(
            !LaneMaxSessions {
                plan: 0,
                implement: 1,
                review: 1,
            }
            .is_valid()
        );
        assert!(
            !LaneMaxSessions {
                plan: 1,
                implement: 11,
                review: 1,
            }
            .is_valid()
        );
    }
}
